<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid source docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="black"
VLINK="black"
ALINK="black">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red>CookingGenes.c</font>
</td>
<td align=right>
<font face=courier size=3>geneid v 1.1</font>
<font face=arial black size=3>source documentation</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Processing of predicted genes in order to extract more information
(gene score, number of exons, ...) about every separarated gene out of
a multiple gene list of linked exons. Depending on the output format,
protein product or genomic sequence of exons might be displayed.
</font>
</div>
</td>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Briefing:</font></td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
void printProt(char* Name,
               long ngen,
               char* prot,
               long nAA,
               int mode)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Display a biological sequence in fasta format (FASTALINE chars per line),
including a previously predefined header starting with ">".
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
void selectFeatures(char* exonType,
                    char exonStrand,
                    profile** p1,
                    profile** p2,
                    int* type1,
                    int* type2,
                    int* strand,
                    gparam* gp)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Given an exon and its list of properties, this function returns: the type
of its two signals including their profile, and its strand. (useful to print
in extend format)
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
long CookingInfo(exonGFF* eorig,
                 gene info[],
                 long* nvExons)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Post-processing of the best genes (geneid format). In geneid, genes are list of exons
linked from the end of the gene to the beginning. Therefore, from the
last exon of the last gene, recursively (bottom-up), predicted genes
are recovered and separated. Thus, forward genes are in this way
-- (BOTTOM) Terminal > Internal >... > First (TOP) -- and reverse genes
in this other -- (BOTTOM) First > Internal > ... >Terminal (TOP) -- At the
same time, information about gene score, first and last exon and number
of exons from the current gene are recorded while pointer jumping. Moreover,
number of annotations used into the final gene prediction is returned
(nvExons). The last exon found must be GHOST exon, an artificial exon
(marked as strand *) meaning end of prediction. Returns the number of genes
found into the input list of exons (based on the standard set of features
provided together with the gene model).
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
void PrintGene(exonGFF* start,
               exonGFF* end,
               char Name[],
               char* s,
               gparam* gp,
               dict* dAA,
               long igen,
               long nAA,
               int** tAA,
               int nExon,
               int nExons)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Print the input gene recursively (in the reversed order), by printing every
exon in the selected format. Extended format is implemented here (printing
for every exon, complete information about its signals in adition to
the exon) by calling the routine SelectFeatures to get the types of signals.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
void CookingGenes(exonGFF *e,
                  char Name[],
                  char* s,
                  gparam* gp,
                  dict* dAA)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Management of post-processing genes to pretty-printing (geneid / XML format)
that includes protein and genomic sequences displayed in fasta format. gff
format does not include any sequence to display.
</font>
</div>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2001<br><p><br><p>
</center>
</body>
</html>




