<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid source docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="black"
VLINK="black"
ALINK="black">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red>geneid.c</font>
</td>
<td align=right>
<font face=courier size=3>geneid v 1.2</font>
<font face=arial black size=3>source documentation</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Main program: management and control of geneid data flow.
</font>
</div>
</td>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Briefing:</font></td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
void main (int argc, char *argv[])
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
geneid main program actions:
<ul>
<li> Initialize accounting structure
<li> Read setup options from command line
<li> Set ratios (for sites/exons memory allocation) from file size estimation
<li> Request memory for main geneid data structures
<li> Read parameters file (statistical model to predict genomic elements)
<li> [optional]. Read annotations and homology SRs from input file 
<li> Read first DNA sequence from input file (more than one allowed)
<li> Processing sequence: upper case, reverse and complement
<li> Divide sequence into overlapped fragments (length = LENGTHSi) 
<li> For every fragment do:
<ul>
<li> Measure G+C content to select the proper isochore 
<li> Forward strand predictions: sites, exons (both scored and filtered)
<li> Reverse strand predictions: sites, exons (both scored and filtered)
<li> Merge forward and reverse predicted exons (and annotations)
<li> Integrate external information: evidence and HSPs
<li> Sort by left (minor) position the whole set of predicted exons
<li> Print genomic elements: sites and exons (selected format)
<li> Run genamic algorithm to assemble the best gene possible
<li> Save necessary information to go on computing the next fragment
<li> Delete the rest of information (to save memory)
</ul>
<li> Print best prediction: multiple gene output (selected format)
</ul>
<br>
There is a partial prediction mode: reading exons directly from input file
and running assembling algorithm (genamic) to display the best gene predicted
</font>
</div>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




