<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid source docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="black"
VLINK="black"
ALINK="black">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red>geneid.h</font>
</td>
<td align=right>
<font face=courier size=3>geneid v 1.2</font>
<font face=arial black size=3>source documentation</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=600 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Definitions of constants, data types and headers of functions used outside
the module which contains the implementation.
</font>
</div>
</td>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Briefing:</font></td>
<tr>
<td bgcolor=skyblue>

<table cellpadding=15>
<tr><td><div class="bulk"><font face="arial black" size=2>
LENGTHSi
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Input sequence is divided into fragments having this length.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
OVERLAP
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Nucleotides shared between 2 consecutive fragments.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RSITES
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Given a DNA sequence, one signal per RSITES nucleotides is
estimated to be predicted on (computing NUMSITES).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
REXONS
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Given a DNA sequence, one exon per REXONS nucleotides is estimated
to be predicted on (computing NUMEXONS).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RBSITES
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Given a DNA sequence, one signal per RBSITES nucleotides is the estimated
ratio of sites which must be copied between 2 fragments
(computing BACKUPSITES).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RBEXONS
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Given a DNA sequence, one exon per RBEXONS nucleotides is the estimated
ratio of exons which must be copied between 2 fragments
(computing BACKUPEXONS).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RFIRST
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
NUMEXONS / RFIRST initial exons are supposed to be predicted in LENGTHSi bases.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RINTER
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
NUMEXONS / RINTER internal exons are supposed to be predicted in LENGTHSi bases.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RTERMI
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
NUMEXONS / RTERMI terminal exons are supposed to be predicted in LENGTHSi bases.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RSINGL
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
NUMEXONS / RSINGL single genes are supposed to be predicted in LENGTHSi bases.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
RORF
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
NUMEXONS / RSINGL ORFs are supposed to be predicted in LENGTHSi bases.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
FSORT
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
FSORT * NUMEXONS is the maximum number of predicted exons (both strands)
in LENGTHSi bases.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
NUMEVIDENCES
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number of annotations per locus 
read from optional file.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXHSP
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number of HSPs (protein homology) read
from optional file (per locus, strand and frame).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXNSEQUENCES
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number of locus in multi-fasta files.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXGENE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number if predicted genes.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXEXONGENE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum amount of exons in every gene.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXAA
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum length of predicted proteins (in amino acids).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXCDNA
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum length of exonic sequence for a gene (cDNA).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXISOCHORES
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number of isochores in the parameters file.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
EXONLENGTH, SINGLEGENELENGTH, ORFLENGTH
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Minimum allowed size for internal exons, single genes and ORFs.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MINEXONLENGTH, MINSCORELENGTH
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Minimum size for exons to compute protein coding potential (score if not).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
ISOCONTEXT
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Size of region around an exon when its G+C content is computed.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
NULL_OLIGO_SCORE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Penalty for N's in scoring exons.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
LOCUSLENGTH
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number of chars for the name of the input sequence.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
OLIGOLENGTH
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum length for the oligonucleotides used in Markov chains
(scoring sites and exons).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
VERSION
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Current geneid version.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
SITES, EXONS, EVIDENCE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Field feature in gff standard.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
FRAMES, STRANDS, FORWARD, REVERSE, ACC, DON, STA, STO, FIRST,
INTERNAL, TERMINAL, SINGLE, ORF, LENGTHCODON, PERCENT, MEGABYTE,
MAXTIMES, PROT, DNA, MINUTE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Numerical constants.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
sFORWARD, sREVERSE, sACC, sDON, sSTA, sSTO, sFIRST, sINTERNAL,
sTERMINAL, sSINGLE, sORF, sEXON, xmlFORWARD, xmlREVERSE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
String constants.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
BLOCK, NONBLOCK
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Mark gene model rules as blocking or not blocking. To preserve or not
the group structure in gene annotations (evidences).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
COFFSET
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Correction because of arrays in C starts from 0 to N -1.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXLINE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum number of characters read from input line.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXSTRING
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum allowed length of geneid messages and strings.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXENTRY, MAXTYPE, MAXINFO, NOTFOUND
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Dictionary (hash table) definitions.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
HASHFACTOR
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Computing size of hash table to save best genes (sites, exons).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
PARAMETERFILE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Default parameters file: filename.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
FILENAMELENGTH
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Maximum size for filenames.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
INFI, sINFI, INF
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Representing the infinity value.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MAXSCORE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Default value for annotations forced to appear in the final prediction.
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
FASTALINE
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Lenght (maximum) for fasta lines
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
MESSAGE_FREQ
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
Display message with amount of sequence read (frequency).
</font></div></td>
<tr><td><div class="bulk"><font face="arial black" size=2>
NOGROUP
</font></div></td>
<td><div class="bulk"><font face="arial black" size=2>
String to represent (ab initio) exons without group (field 9 in gff).
</font></div></td>
</table>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




