<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid source docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="black"
VLINK="black"
ALINK="black">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red>ReadExonsGFF.c</font>
</td>
<td align=right>
<font face=courier size=3>geneid v 1.2</font>
<font face=arial black size=3>source documentation</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Read an external file of exons. These annotations will be added in the
final gene prediction either forcing to include them or mixing with the
ab initio predictions (depending on the existence of value in the group
field (gff) of annotations to preserve or not the complete annotated gene).
<br><p>
<tt>GFF</tt> format line (<i>tab</i> as field separator):
<ul>
<li>Name: locusname of the input sequence.
<li>Source: program used to produce the output.
<li>Type: biological feature described here.
<li>Begin: left position in the sequence.
<li>End right position in the sequence.
<li><font color=red>Score</font>: probability to be a real prediction.
<li>Strand: forward or reverse reading sense.
<li><font color=red>Frame</font>: number of nucleotides of the left non complete codon.
<li><font color=red>[group]</font>: The way to group a set of exons forming a gene. 
</ul>
<font color="green">
<tt>geneid</tt> wildcards: represented with a dot '<font color=green size=+2>.</font>'
<ul>
<li>Score: maximum score will be assigned to the annotation.
<li>Frame: three exons will be created from this one (3 reading frames).
<li>Group: If is present, the set of annotations (exons) with the same group
will be preserved as a complete gene as long as they match the gene assembling
rules. If not, annotations will be mixed with ab initio predictions as long
as they might contribute to be part of the best genes.
</ul>
</font>
<center><font color="red">-- Annotations MUST be ordered by left position (beginning) --</font></center>
</font>
</div>
</td>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Briefing:</font></td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
long ReadExonsGFF (char *FileName,
                   packEvidence* pv,
                   dict* d)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Input exons from an external file. They must be sorted by the left position
(increasing order). It returns the number of created annotations. Exons
with unknown (not in gene model) features are skipped, displaying one warning
for every one.
</font>
</div>
</td>
<tr>
<td><font face=courier size=2 color=red>
<pre>
packEvidence* SelectEvidence(packExternalInformation* external, 
                             char* Locus)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Select the group of annotations to be integrated into the predictions according 
to current Locus name (using a hash table - dictionary of locus). 
</font>
</div>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




