<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid source docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="black"
VLINK="black"
ALINK="black">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red>RequestMemory.c</font>
</td>
<td align=right>
<font face=courier size=3>geneid v 1.2</font>
<font face=arial black size=3>source documentation</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
This set of functions are the responsible of memory management, allocating
main geneid data structures, from predictions to parameters and statistical
models, as well as temporary structures reset in every fragment of the
input sequence.
</font>
</div>
</td>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Briefing:</font></td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
account* RequestMemoryAccounting()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for account data structure.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
char* RequestMemorySequence(long L)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for a DNA sequence.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packSites* RequestMemorySites()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for main structure pack of sites (forward or reverse sense)
and every type of signal: acceptor and donor splice sites, start and
stop in translation.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packExons* RequestMemoryExons()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for pack of exons and every type of exons:
NUMEXONS is modified (divided) by RSINGL, RFIRST, RINTER, RTERMI and RORF
ratios because some type of exons (internal and terminal) appear more
frequently than others.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
exonGFF* RequestMemorySortExons()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory to sort exons found in the current fragment.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packEvidence* RequestMemoryEvidence()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for input annotations (evidences).
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packHSP* RequestMemoryHomology()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for input similarity to protein regions (homology).
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packExternalInformation* RequestMemoryExternalInformation()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for external information (dictionary of Locus names)
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packGC* RequestMemoryGC()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for the precomputed array of G/C's and N's nucleotides on a
DNA subsequence.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
gparam* RequestMemoryParams()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for the statistical model of one isochore loading from
the parameter file: profiles, Markov initial/transition arrays, exon scoring
parameters. Memory for temporary arrays which compute accumulated sum of
penta/hexanucleotides in every split.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
gparam ** RequestMemoryIsochoresParams()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for the array of isochores and their common information:
gene model dictionary and allowed distances arrays.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
void RequestMemoryProfile(profile* p)
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for every position containing a Markov chain inside in a Position
Weight Array (profile).
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packGenes* RequestMemoryGenes()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for the main structure, Ghost exon, best partial genes (Ga) and
sorting by donor functions (d-array) in every gene class, necessary in
genamic.
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
packDump* RequestMemoryDumpster()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for backup information between two fragments: sites, exons
and a hash table useful to save only once every exon and avoid
usual redundancy between best partial genes (they share most exons).
</font>
</div>
</td>
<tr>
<td><font face=courier size=4 color=red>
<pre>
dict* RequestMemoryAaDictionary()
</pre>
</font>
</td>
<tr>
<td bgcolor=skyblue>
<div class="bulk"><font face="arial black" size=2>
Memory for the genetic code, a hash table to translate from predicted genes
or exons into proteins.
</font>
</div>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




