<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="blue"
VLINK="blue"
ALINK="blue">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red><b>geneid documentation:</b></font>
</td>
<td align=right>
<font face="arial black" size=3>2. Signals, exons and genes</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor="red"><font face="arial black" size=3>Table of contents:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<ol>
<li> <a href="#context">General context</a>
<li> <a href="#signals">Signals</a>
<li> <a href="#frames">Frame and remainder</a>
<li> <a href="#exons">Exons</a>
<li> <a href="#genes">Genes</a>
<li> <a href="./../index.html">INDEX</a>
</ol>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>General context:</font>
<a name ="context"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<tt>geneid</tt> is a program to predict genes in anonymous genomic sequences
designed with a hierarchical structure. First of all, splice sites and
start/stop codons are predicted and scored along the DNA sequence by using
Position Weight Arrays (PWA). Secondly, exons are built from the set of
predicted signals before. Exons are scored as the sum of the scores
of the defining sites, plus the log-likelihood ratio of a Markov model
to distinguish between coding and non-coding DNA. Finally, from the
set of predicted exons, the best gene structure is assembled maximizing the
sum of the scores of the assembled exons.
</font>
</div>
</td>
</table>


<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Signals:</font>
<a name ="signals"></td>
<tr>
<td>
<font face="arial black" size=2>
<tt>geneid</tt> is able to detect the following signals:
<ul>
<li> <b>Acceptor splice sites</b>:
<i>Last intron dinucleotide mandatory <tt>AG</tt></i>.<br>
Acceptors occur at the limits between intron/exon.
<br>
The following signals associated to the acceptors are also detected (optional):
<ol>
<li> Branch point
<li> Poly Pyrimidine Tract
</ol>
<li> <b>Donor splice sites</b>:
<i>First intron dinucleotide mandatory <tt>GT</tt></i>.<br>
Donors occur at the limits between exon/intron.
<li> <b>Start codons (translation)</b>:
<i>First exon first trinucleotide mandatory <tt>ATG</tt></i>.<br>
Initiation of transcription from mRNA into protein.
<li> <b>Stop codons (translation)</b>:
<i>Three possible codons <tt>TAA, TAG and TGA</tt></i>.<br>
End of transcription from mRNA into protein.
</ul>
</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Frame and remainder:</font>
<a name ="frames"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<tt>geneid</tt> exons are Open Reading Frames delimitited by signals. Frame
and remainder are 2 values assigned to exons which have to be understood in
the following way:
<br><p>
<b>FRAME:</b>
The number of nucleotides (0,1,2) from the first nucleotide in the exon sequence
to the the first nucleotide in the first COMPLETE codon translated from the
genomic sequence.

<br><p>

<b>REMAINDER:</b>
The number of nucleotides left (0,1,2) after the last COMPLETE codon has
been defined from the exon sequence, given its frame. Therefore, exon
remainder is completely determined by the length of the exon and its frame.
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Exons:</font>
<a name ="exons"></td>
<tr>
<td>
<font face="arial black" size=2>
<tt>geneid</tt> is able to build the following type of exons:
<ul>
<li> <b>First exons</b>:
<i>Delimited by Start and Donor</i>.<br>
Frame = 0 - Remainder = 0,1,2
<li> <b>Internal exons</b>:
<i>Delimited by Acceptor and Donor</i>.<br>
Frame = 0,1,2 - Remainder = 0,1,2
<li> <b>Terminal exons</b>:
<i>Delimited by Acceptor and Stop</i>.<br>
Frame = 0,1,2 - Remainder = 0
<li> <b>Single genes</b>:
<i>Delimited by Start and Stop</i>.<br>
Frame = 0 - Remainder = 0
</ul>
<font color="red">NOTE: The character 'X' denotes unknown codons in the 
translation process<br>(such those cases in which there are masked 
nucleotides, "N", within the codon sequence).</font>
</font>
</td>
</table>


<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Genes:</font>
<a name ="genes"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Gene structures are series of frame compatible exons that are assembled
following some fixed rules and restrictions specified in a Gene Model.
Two exons are frame compatible if the frame of the second exon plus
the remainder of the first one MODULUS 3 is 0. Predicted exons are connected
as long as frame compatibility rule and gene model restrictions are
respected properly.
</font>
</div>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




