<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="blue"
VLINK="blue"
ALINK="blue">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red><b>geneid documentation:</b></font>
</td>
<td align=right>
<font face="arial black" size=3>4. Running <tt>geneid</tt>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor="red"><font face="arial black" size=3>Table of contents:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<ol>
<li> <a href="#fasta">FASTA format</a>
<li> <a href="#main">General considerations</a>
<li> <a href="#menu"><tt>geneid</tt> command line options</a>
<li> <a href="#output"><tt>geneid</tt> output formats</a>
<li> <a href="#samples">Samples</a>
<li> <a href="#multi">Multi-fasta files</a>
<li> <a href="./../index.html">INDEX</a>
</ol>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>FASTA format:</font>
<a name ="fasta"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<tt>geneid</tt> input sequences must be in FASTA format:
<br><p>
FASTA sequence files show this structure:
<br><p>
<center>
<table border=2 cellpadding=10>
<tr><td bgcolor="skyblue">
<pre>
>NAME OR IDENTIFIER OF SEQUENCE
ATCGTACAGCTAGCTAGCTACGTACG
TCGTACGATCGATCGTAGCATCACGA
GGTCAGCATCTAGCACATCACACAGC
CTATCAGCTAGCATCGATCGCATCGA
...
</pre>
</td>
</table>
</center>
<br><p>
Usually Fasta lines are 60 chars long but this is not mandatory.
</font>
</div>
</td>
</table>


<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>General considerations:</font>
<a name ="main"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
To run <tt>geneid</tt>,
type <tt><font color="red">geneid [options] sequence_file</font></tt> where
sequence_file is a FASTA formatted DNA sequence. Predictions are sent to
the standard output. In addition, the program requires an organism specific
parameters file (which can be broadly used within the phylogenetic group to which
organism belongs. For instance, the human file may be used as well to
predict genes in genomic sequences from vertebrates).The directory
<tt>params</tt> contains several files for different organisms.
<br><p>
The parameters file may be speficied in three different ways:
<ol>
<li> By default, <tt>geneid</tt> will try to find a default file called
<tt>param.default</tt> which must be in the same directory that the binary.
<li> Through the enviromental var GENEID, defining the path.
<li> Using the command line option <tt><font color="red">geneid -P file</font></tt>.
</ol>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3><tt>geneid</tt> command line options:</font>
<a name ="menu">
</td>
<tr>
<td align=center>
<table border=0 width=600 cellpadding=10>
<tr>
<td></td>
<tr>
<td bgcolor="lightblue">
<table border=0 width=600>
<tr>
<td width=700 align=right colspan=2><font face="arial" size=4>Selection of genomic
elements to display</td>
<tr>
<td width=50 bgcolor="yellow"><font face="arial" size=4>Flag</td>
<td width=650 bgcolor="yellow"><font face="arial" size=4>Action</td>
<tr>
<td width=50><font face="arial" size=3>b</td>
<td width=650><font face="arial" size=3>Display best Start codons
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>d</td>
<td width=650><font face="arial" size=3>Display best Donor sites
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>a</td>
<td width=650><font face="arial" size=3>Display best Acceptor sites
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>e</td>
<td width=650><font face="arial" size=3>Display best Stop codons
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>f</td>
<td width=650><font face="arial" size=3>Display best First exons
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>i</td>
<td width=650><font face="arial" size=3>Display best Internal exons
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>t</td>
<td width=650><font face="arial" size=3>Display best Terminal exons
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>s</td>
<td width=650><font face="arial" size=3>Display best Single genes
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>x</td>
<td width=650><font face="arial" size=3>Display best exons (all)
found on sequence</td>
<tr>
<td width=50><font face="arial" size=3>z</td>
<td width=650><font face="arial" size=3>Display Open Reading Frames
found on sequence [see -Z]</td>
<tr>
<td width=50><font face="arial" size=3>D</td>
<td width=650><font face="arial" size=3>Display genomic sequence in
predicted genes</td>
<tr>
<td><br><p></td>
<tr>
<td width=700 align=right colspan=2><font face="arial" size=4>Selection of
output format</td>
<tr>
<td width=50 bgcolor="yellow"><font face="arial" size=4>Flag</td>
<td width=650 bgcolor="yellow"><font face="arial" size=4>Action</td>
<tr>
<td width=50><font face="arial" size=3>-</td>
<td width=650><font face="arial" size=3>geneid format</td>
<tr>
<td width=50><font face="arial" size=3>G</td>
<td width=650><font face="arial" size=3>gff format</td>
<tr>
<td width=50><font face="arial" size=3>X</td>
<td width=650><font face="arial" size=3>Extended format (gff or geneid)</td>
<tr>
<td width=50><font face="arial" size=3>M</td>
<td width=650><font face="arial" size=3>XML output (see -m)</td>
<tr>
<td width=50><font face="arial" size=3>m</td>
<td width=650><font face="arial" size=3>See DTD for XML output</td>
<tr>
<td><br><p></td>
<tr>
<td width=700 align=right colspan=2><font face="arial" size=4>
Control prediction engine</td>
<tr>
<td width=50 bgcolor="yellow"><font face="arial" size=4>Flag</td>
<td width=650 bgcolor="yellow"><font face="arial" size=4>Action</td>
<tr>
<td width=50><font face="arial" size=3>-</td>
<td width=650><font face="arial" size=3>Predictions in both strands
of sequence</td>
<tr>
<td width=50><font face="arial" size=3>W</td>
<td width=650><font face="arial" size=3>Predictions only in forward strand
of sequence</td>
<tr>
<td width=50><font face="arial" size=3>C</td>
<td width=650><font face="arial" size=3>Predictions only in reverse strand
of sequence</td>
<tr>
<td width=50><font face="arial" size=3>o</td>
<td width=650><font face="arial" size=3>Only signal and exon prediction (gene
assembling disabled)</td>
<tr>
<td width=50><font face="arial" size=3>F</td>
<td width=650><font face="arial" size=3>Force prediction of a complete gene</td>
<tr>
<td width=50><font face="arial" size=3>O</td>
<td width=650><font face="arial" size=3>Only gene assembling (exons provided
from file)</td>
<tr>
<td width=50><font face="arial" size=3>Z</td>
<td width=650><font face="arial" size=3>Switch ORF searching on</td>
<tr>
<td><br><p></td>
<tr>
<td width=700 align=right colspan=2><font face="arial" size=4>
Re-annotation of sequences
</td>
<tr>
<td width=50 bgcolor="yellow"><font face="arial" size=4>Flag</td>
<td width=650 bgcolor="yellow"><font face="arial" size=4>Action</td>
<tr>
<td width=50><font face="arial" size=3>R</td>
<td width=650><font face="arial" size=3>Include annotations (evidences) provided
from file</td>
<tr>
<td width=50><font face="arial" size=3>S</td>
<td width=650><font face="arial" size=3>Use homology to protein information
 (SR) provided from file</td>
<tr>
<td><br><p></td>
<tr>
<td width=700 align=right colspan=2><font face="arial" size=4>
Statistical model
</td>
<tr>
<td width=50 bgcolor="yellow"><font face="arial" size=4>Flag</td>
<td width=650 bgcolor="yellow"><font face="arial" size=4>Action</td>
<tr>
<td width=50><font face="arial" size=3>P</td>
<td width=650><font face="arial" size=3>Provide a new parameter file</td>
<tr>
<td width=50><font face="arial" size=3>E</td>
<td width=650><font face="arial" size=3>Increase/decrease exon weight
value</td>
<tr>
<td><br><p></td>
<tr>
<td width=700 align=right colspan=2><font face="arial" size=4>
Miscellanea
</td>
<tr>
<td width=50 bgcolor="yellow"><font face="arial" size=4>Flag</td>
<td width=650 bgcolor="yellow"><font face="arial" size=4>Action</td>
<tr>
<td width=50><font face="arial" size=3>v</td>
<td width=650><font face="arial" size=3>Display information during geneid
running</td>
<tr>
<td width=50><font face="arial" size=3>B</td>
<td width=650><font face="arial" size=3>Show memory requirements</td>
<tr>
<td width=50><font face="arial" size=3>h</td>
<td width=650><font face="arial" size=3>Show list of available options</td>
</table>
</table>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3><tt>geneid</tt> output formats:</font>
<a name ="output"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Click over every available format to see a detailed description:
<ul>
<li><a href="./formats_html/geneid.html">geneid format</a> (default)
<li><a href="./formats_html/gff.html">gff format</a>
<li><a href="./formats_html/xml.html">xml format</a>
<li><a href="./formats_html/extended.html">extended format (geneid and gff formats)</a>
</ul>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Samples:</font>
<a name ="samples"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Click over every example to see the corresponding output:
<ul>
<li> <a href="./samples_html/acceptors.html">Prediction of acceptor splice sites</a>
<li> <a href="./samples_html/exons.html">Prediction of exons</a>
<li> <a href="./samples_html/genes.html">Gene prediction</a>
<li> <a href="./samples_html/evidences.html">External information 1: annotations</a>
<li> <a href="./samples_html/sr.html">External information 2: sequence homology</a>
</ul>
<font color="red">This the sequence used to generate these outputs: <a href="./sequence/test.fa">test.fa</a></font>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Multi-fasta files:</font>
<a name ="multi"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<tt>geneid</tt> can process files containing more than one FASTA sequence. 
Moreover, external information as annotations and HSPs can be provided for 
every sequence. <b>Requirement</b>: records belonging to the same Locus must be sorted 
by first position.<br><p>

Click over every example to see the corresponding output:
<ul>
<li> <a href="./sequence/5seqs.fa">Multi-fasta file</a>
<li> <a href="./sequence/tmp.evi">Multi-fasta annotations</a>
<li> <a href="./sequence/tmp.hsp">Multi-fasta homology information</a>
</ul>
</font>
</div>
</td>
</table>



<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




