<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="blue"
VLINK="blue"
ALINK="blue">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red><b>geneid documentation:</b></font>
</td>
<td align=right>
<font face="arial black" size=3>6. Introducing external information: annotations</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor="red"><font face="arial black" size=3>Table of contents:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<ol>
<li> <a href="#description">Description</a>
<li> <a href="./../index.html">INDEX</a>
</ol>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font>
<a name ="description"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<tt>geneid</tt> allows to integrate ab initio predictions with external
evidence (annotations), such as annotated genes. The external
evidence is read from an additional file in <tt>gff</tt> format. <tt>geneid</tt>
is instructed to read this file with the command line option <tt>-R filename</tt>.
This input file must sorted by the starting position (column 4 in <tt>gff</tt>).
<br><p>
There is a difference between using options <tt>-O filename</tt> and
<tt>-R filename</tt>. In the first case, only elements extracted from file
will be assembled while, in the second one, gene predictions are built from
both file records and from <tt>geneid</tt> predictions.
<br><p>
If the elements in the input file are assigned an score, then they will
"compete" with <tt>geneid</tt> original predictions (if any) to be in the final gene
structure. For instance, this record will fight against ab initio predictions:
<br><p>
<center>
<table border=1 cellpadding=10 width=700>
<tr>
<td align=center bgcolor="skyblue">
<pre>
AE002566   lab_XX  Internal    66255    66323    3.27  +  1
</pre>
</td>
</table>
</center>
<br><p>
If no score is given for the element (a dot "."), then this element is
supposed to be mandatory (forced) in the final gene prediction (unless a conflicting
element with no score is also given in the input file). For instance, this record
will be in the final prediction.
<br><p>
<center>
<table border=1 cellpadding=10 width=700>
<tr>
<td align=center bgcolor="skyblue">
<pre>
AE002566   lab_XX  Internal    66255    66323    .  +  1
</pre>
</td>
</table>
</center>
<br><p>
The frame can be either set in the column 8 of <tt>gff</tt> format or
skipped by using the wildcar "." when is unknown. In the last case,
<tt>geneid</tt> generates 3 equivalent elements (one per possible frame),
computing the corresponding remainder in each case, keeping the frame/remainder
consistency anyway when assembling is done. For instance, this record is
internally expanded to these 3 exons, being incorporated to the set of
candidate exons to be part of final gene prediction:
<br><p>
<center>
<table border=1 cellpadding=10 width=700>
<tr>
<td align=center bgcolor="skyblue">
<pre>
AE002566   lab_XX  Internal    66255    66323    3.27  +  .
<hr>
AE002566   lab_XX  Internal    66255    66323    3.27  +  0
AE002566   lab_XX  Internal    66255    66323    3.27  +  1
AE002566   lab_XX  Internal    66255    66323    3.27  +  2
</pre>
</td>
</table>
</center>
<br><p>
By using the optional group field (column 9 in <tt>gff</tt> format), user
is able to specificy whether one annotated gene (annotation) introduced
in the input file has to be preserved if it is incorporated in the final results
or <tt>geneid</tt> predictions can be mixed within that annotation.
For instance, given this annotation (1), this gene will be preserved
in the final output. But, given this other annotation (2) for the same gene
but without setting a group identifier, we can obtain predictions such as
the following (3):
<br><p>
<center>
<table border=1 cellpadding=10 width=700>
<tr>
<td align=left bgcolor="skyblue">
<pre>
(1)
AE002566  external Terminal  21839    22922   18.37 -  1  gene_2
AE002566  external Internal  23679    24029    7.99 -  1  gene_2
AE002566  external First     30732    30775   -1.11 -  0  gene_2
<hr>
(2)
AE002566  external Terminal  21839    22922   18.37 -  1  
AE002566  external Internal  23679    24029    7.99 -  1  
AE002566  external First     30732    30775   -1.11 -  0  
<hr>
(3)
AE002566     external  Terminal  21839    22922   18.37 -  1  
AE002566     external  Internal  23679    24029    7.99 -  1  
<font color="red">AE002566  geneid_v1.2  Internal  28002    28007    1.14 -  1 </font>
AE002566     external  First     30732    30775   -1.11 -  0
</pre>
</td>
</table>
</center>







</font>
</div>
</td>
</table>


<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




