<html>
<head>
<style type="text/css">
div.bulk { text-align: justify }
</style>
<title>
geneid docs
</title>
</head>

<BODY 
TEXT color ="blue"
BGCOLOR="white"
LINK="blue"
VLINK="blue"
ALINK="blue">

<center>
<table border=0 width=700>
<tr>
<td align=left>
<font face=courier size=5 color=red><b>geneid documentation:</b></font>
</td>
<td align=right>
<font face="arial black" size=3>7. Using sequence homology information</font>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor="red"><font face="arial black" size=3>Table of contents:</font></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<ol>
<li> <a href="#description">Description</a>
<li> <a href="#srs">HSPs</a>
<li> <a href="#frame">Frame definition in <tt>blast</tt> and <tt>geneid</tt></a>
<li> <a href="./../index.html">INDEX</a>
</ol>
</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>Description:</font>
<a name ="description"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<tt>geneid</tt> provides a simple support to the usage of homology
information, either genomic homology (DNA - DNA) or protein homology
(Protein - DNA translated). Homology information extracted from the aligment
between input sequence and others (well-annotated) can be used to improve
the degree of accuracy obtained ab initio, increasing the final score on the
set of predicted exons overlapping the shared regions in such alignment
(High-scoring Segment Pairs).
<br><p>
The information from aligments is read from an additional file in
<tt>gff</tt> format as a list of HSPs. <tt>geneid</tt>
is instructed to read this file with the command line option 
<tt>-S filename</tt>. It is NOT necessary to provide the HSPs ordered 
by any position in the sequence.
<br><p>
The frame can be either set in the column 8 of <tt>gff</tt> format or
skipped by using the wildcar "." when is unknown. In the last case,
<tt>geneid</tt> generates 3 equivalent elements (one per possible frame).
This is necessary when aligments have been produced between 2 genomic
sequences (DNA to DNA) and therefore there is no frame associated to.
<br><p>
The group field (column 9 in <tt>gff</tt> format) can be present but will be
omitted when storing this information.
</font>
</div>
</td>
</table>

<br><p>

<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>HSPs:</font>
<a name ="srs"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
Given an <tt>blast</tt> aligment, a set of High-score Segment Pairs (HSPs)
are produced to represent parts common in both sequences. Given the
interesting sequence (input to <tt>geneid</tt>), numerous pairwise
alignments can be done by using sequences related to the original by
many causes such as containing an homologous gene (syntenic gene prediction).
<br><p>
<tt>geneid</tt> will increase proportionally the score of predicted exons
which overlap the HSPs obtained by previous pairwise aligments. To sum up
the information coming from several aligments, a projection of HSPs over the
sequence is performed, taking always the highest score among the HSPs 
being in every position of the input sequence. Exon positions overlapping
regions without homology support will be given a value <tt>NO_SCORE</tt> 
(see <a href="../chapter8/index.html">chapter 8. <tt>geneid</tt> 
parameter file</a>)
<br><p>
<center>
<IMG SRC="../images/SR.jpg">
</center>

</font>
</div>
</td>
</table>

<br><p><br><p>
<table border=0 width=700 cellpadding=10>
<tr>
<td bgcolor=orange><font face="arial black" size=3>
Frame definition in <tt>blast</tt> and <tt>geneid</tt>
</font>
<a name ="frame"></td>
<tr>
<td>
<div class="bulk"><font face="arial black" size=2>
<font color="red">
It is important to understand that frame definition in <tt>blast</tt> and
<tt>geneid</tt> is different and therefore a simple translation from one into
another format is needed. Altough this conversion is implemented inside
<tt>geneid</tt>, it might be necessary to keep it in mind to analyze and
understand the new results.
</font>
<br><p>
<hr>
<br><p>
In <tt>blast</tt>, there are 3 frames (1,2 or 3)
corresponding to the mathematical modulus operation computed following this
formula:
<br><p>
<center>
frame = HSP starting position modulus 3 (0 corresponds to frame 3).
</center>
<br><p>
i.e. Given a sequence which its first position is called 1, one HSP which starts
at the position 7 is said to be in frame 1 (7 modulus 3 = 1), while a HSP
which starts at the position 9 is said to be in frame 3 (9 modulus 3 = 0
and 0 is coded by 3).
<br><p>
<font color="green">
IMPORTANT: HSPs might appear in the negative (reverse) strand of the
input sequence. In that case, the HSPs coordinates must be translated into
coordinates regarding to the reverse reading direction before computing the
frame.
</font>
<br><p>
<hr>
<br><p>
In <tt>geneid</tt>, the frame is defined as the number of nucleotides (0,1,2)
from the first nucleotide in the exon sequence to the the first nucleotide
in the first COMPLETE codon translated from the genomic sequence
(<a href="../chapter2/index.html#frames">see section Frame and Remainder (chapter 2)</a>).
<br><p>
<hr>
<br><p>
Obviously, it is necessary a method to convert <tt>geneid</tt> frames (exons)
into <tt>blast</tt> frames (HSPs) to use this information properly. The
following formula makes both types of frame definitions compatible
<br><p>
  Given an exon and a HSP (both in the same strand):
<center>
<br><p>
  blastFrame(HSP) = (starting(exon) + geneidFrame(exon)) % 3;<br>
  (result 0 corresponding to blastFrame 3)
</center>
<br><p>
  notice: blastFrame = [1,2,3] and geneidFrame = [0,1,2].
</font>
</div>
</td>
</table>

<br><p><br><p><br><p>

<table border=0>
<tr>
<td align=left bgcolor=darkgreen>
<script>
var modifieddate=document.lastModified
    document.write("<font size=1 face  = 'arial black'  color='white'>" + 
                   "Last modified: " + modifieddate)
</script>
</td>
</table>

<br><p>
Enrique Blanco Garcia &copy; 2003<br><p><br><p>
</center>
</body>
</html>




